src/maturin/springs.clj

```diff
@@ -0,0 +1,62 @@
+;; create xs
+(defn gen-xs [n]
+    (take n (repeatedly #(rand 10))))
+  
+;; create linear regression mock data
+(defn gen-ys [xs slope intercept]
+    (map #(+ (d/sample (d/normal 0 1)) intercept (* % slope)) xs))
+
+;; return linear function
+(defn linear-func [slope intercept]
+    (fn [x] (+ intercept (* slope x)))) 
+
+;; lagrangian for the springs-rod system, regression case:
+;; springs constrained to have the same x-coordiate on both axes = minimize MSE
+(defn L-regression
+    [xs ys k]
+    (fn [[_ [slope intercept] _]]
+        (let [preds (map (linear-func slope intercept) xs)
+            diffs (mapv - preds ys)]
+            (- (* 1/2 k (square diffs))))))
+
+(defn distance-point-line [slope intercept]
+    (fn [x y]
+        (/ (Math/abs (+ (* slope x) (- y) intercept)) (Math/sqrt (square [slope 1])))))
+
+;; lagrangian for the springs-rod system, PCA case:
+;; springs can "slide" on the bar's end = minimize point-line distance
+(defn L-PCA
+    [xs ys k]
+    (fn [[_ [slope intercept] _]]
+        (let [diffs (mapv (distance-point-line slope intercept) xs ys)]            
+            (- (* 1/2 k (square diffs))))))
+        
+
+;; trying to draw _something_, prettify later
+(defn draw-line [convert]
+(fn [{:keys [state color]}]
+    (q/background 100)
+    (q/fill color 255 255)
+    (let [[slope intercept] (convert state)]
+        (attach [0 intercept] [10 ((linear-func slope intercept) 10)])
```

Comment:
this is very cool. If you take my idea from above, about composing, I think you could use the exact same transformation to get positions for all of the points.

Reply:
I still can't get the graphics to work; the things I'm curious about (but can't experiment with at the moment) - if the drawing function only takes in the state - the intercept and the slope - how can I draw the xs and ys?